airbyte-cdk/java/airbyte-cdk/db-destinations/src/main/java/io/airbyte/cdk/integrations/destination/jdbc/typing_deduping/JdbcSqlGenerator.java

```diff
@@ -83,4 +132,347 @@ protected DataType<?> toDialectType(final AirbyteProtocolType airbyteProtocolTyp
+            .select(with(rawTableRowsWithCast)
+                        .with(filteredRows)
+                        .select(finalTableFields)
+                        .from(filteredRows)
+                        .where(field(ROW_NUMBER_COLUMN_NAME, Integer.class).eq(1)) // Can refer by CTE.field but no use since we don't strongly type them.
+            )
+            .getSQL(ParamType.INLINED);
+
+    // Used for append and overwrite modes.
+    final String insertStmt =
+        insertIntoFinalTable(finalSchema, finalTable, streamConfig.columns(), getFinalTableMetaColumns(true))
+            .select(with(rawTableRowsWithCast)
+                        .select(finalTableFields)
+                        .from(rawTableRowsWithCast))
+            .getSQL(ParamType.INLINED);
+    final String deleteStmt = deleteFromFinalTable(finalSchema, finalTable, streamConfig.primaryKey(), streamConfig.cursor());
+    final String deleteCdcDeletesStmt =
+        streamConfig.columns().containsKey(getCdcDeletedAtColumn()) ? deleteFromFinalTableCdcDeletes(finalSchema, finalTable) : "";
+    final String checkpointStmt = checkpointRawTable(rawSchema, rawTable, minRawTimestamp);
+
+    if (streamConfig.destinationSyncMode() != DestinationSyncMode.APPEND_DEDUP) {
+      return Strings.join(
+          List.of(
+              beginTransaction(),
+              insertStmt,
+              checkpointStmt,
+              commitTransactionInternal()),
+          ";" + System.lineSeparator());
+    }
+
+    // For append-dedupe
+    return Strings.join(
+        List.of(
+            beginTransaction(),
+            insertStmtWithDedupe,
+            deleteStmt,
+            deleteCdcDeletesStmt,
+            checkpointStmt,
+            commitTransactionInternal()),
+        ";" + System.lineSeparator());
+  }
+
+  private String mergeTransaction(final StreamConfig streamConfig,
+                                  final String finalSuffix,
+                                  final Optional<Instant> minRawTimestamp,
+                                  final boolean useExpensiveSaferCasting) {
+
+    throw new UnsupportedOperationException("Not implemented yet");
+
+  }
+
+  protected String createSchemaSql(final String namespace) {
+    DSLContext dsl = getDslContext();
+    CreateSchemaFinalStep createSchemaSql = dsl.createSchemaIfNotExists(quotedName(namespace));
+    return createSchemaSql.getSQL();
+  }
+
+  protected String createTableSql(final String namespace, final String tableName, final LinkedHashMap<ColumnId, AirbyteType> columns) {
+    DSLContext dsl = getDslContext();
+    CreateTableColumnStep createTableSql = dsl
+        .createTable(quotedName(namespace, tableName))
+        .columns(buildFinalTableFields(columns, getFinalTableMetaColumns(true)));;
+    return createTableSql.getSQL();
+  }
+
+  protected String beginTransaction() {
+    return "BEGIN";
```

Comment:
I just learned that jooq supports transactions out of the box, if you can pass around a `Configuration` object (https://www.jooq.org/doc/latest/manual/sql-execution/transaction-management/) - would that help clean up some of our code?

(it might also reduce our need to manually concatenate sql statements together >.>)

Reply:
Yes it does, I was irritated with this last semicolon missing after COMMIT problem too. The challenge was (when doing redshift) didn't wan't to radically change the execution layer. but `DSLContext.transaction(stmt::execute)` is the only way to run it and I couldnt' figure how to extract sql (`toString()` isn't great idea)

Reply:
aaaahhhh I didn't notice that this only works if you use jooq execution. welp. I guess this is how we live then :shrug: